### PR TITLE
hasjob PWA

### PR DESCRIPTION
**Pull request by vidya-ram**
1. Introduced Webpack to build & bundle hasjob's css & js
2. Used Workbox Webpack plugin to generate service worker that takes care of caching & offline support
3. Add Manifest.json to enable "Add to homescreen" feature.
4. Add the new Hasjob logo
add to homescreen&nbsp;&nbsp;&nbsp;icon on homescreen

### CODE REVIEW

File: hasjob/templates/layout.html.jinja2 [hunk 1]
```diff
@@ -229,5 +240,20 @@
...
 {% block layoutscripts %}
   {%- if header_campaign %}{{ campaign_script() }}{% endif %}
+  <script src='{{ asset_path('manifest') | safe }}' type="text/javascript"></script>
+  <script src='{{ asset_path('vendor') | safe }}' type="text/javascript"></script>
+  <script src='{{ asset_path('app') | safe }}' type="text/javascript"></script>
+  <script type="text/javascript">
+    if ('serviceWorker' in navigator) {
+      navigator.serviceWorker.register('/service-worker.js')
+      .then(function(registration) {
+        console.log('Service Worker registration successful with scope: ',
```
Comment (shreyas-satish) [inline hasjob/templates/layout.html.jinja2]: Perhaps better to raise an exception with Sentry?

**Comment by iambibhas** [comment]
I really wish we didn't include built static files in the repo. Can we move the building part to the server? just include the main static files in repo and then when deploying, run build command.

File: hasjob/__init__.py [hunk 2]
```diff
@@ -47,9 +45,8 @@
 
 RQ(app)
 
-baseframe.init_app(app, requires=['hasjob'],
-    ext_requires=['baseframe-bs3',
-        ('jquery.autosize', 'jquery.sparkline', 'jquery.liblink', 'jquery.wnumb', 'jquery.nouislider'),
+baseframe.init_app(app, requires=['baseframe-bs3',
+        'jquery.autosize', 'jquery.sparkline', 'jquery.liblink', 'jquery.wnumb', 'jquery.nouislider',
```
Comment (jace) [inline hasjob/__init__.py]: We stopped using `sparkline` a while ago. Safe to remove now.

File: hasjob/static/js/styles-app.js [hunk 3]
```diff
@@ -0,0 +1 @@
+import '../sass/app.sass';
```
Comment (jace) [inline hasjob/static/js/styles-app.js]: If this needs compilation, it should be in the `assets` folder, not `static`.

Comment (jace) [reply hasjob/static/js/styles-app.js]: Everything in `static` is published to the world and can be directly accessed from a URL.

File: hasjob/static/sass/_header.sass [hunk 4]
```diff
@@ -1,6 +1,6 @@
 body
   font-family: $font-default
-  background: #e8e7e6 image-url("canvas.jpg")
+  background: #e8e7e6 url("~/static/img/canvas.jpg")
```
Comment (jace) [inline hasjob/static/sass/_header.sass]: We should move SASS files into `assets` as well, which means this path will need to be updated (image files stay in `static`).

File: hasjob/templates/admin_dashboard.html.jinja2 [hunk 5]
```diff
@@ -12,7 +12,7 @@
   </div>
 {% endblock %}
 
-{% block footerscripts %}
+{% block tablelayoutscripts %}
```
Comment (jace) [inline hasjob/templates/admin_dashboard.html.jinja2]: This should be `tablayout`, not `tablelayout`. Please amend in the source template as well.

File: hasjob/templates/layout.html.jinja2 [hunk 6]
```diff
@@ -229,5 +240,20 @@
 
 {% block layoutscripts %}
   {%- if header_campaign %}{{ campaign_script() }}{% endif %}
+  <script src='{{ asset_path('manifest') | safe }}' type="text/javascript"></script>
+  <script src='{{ asset_path('vendor') | safe }}' type="text/javascript"></script>
+  <script src='{{ asset_path('app') | safe }}' type="text/javascript"></script>
```
Comment (jace) [inline hasjob/templates/layout.html.jinja2]: Not safe to use the `| safe` filter. If `asset_path` is confirming the strings are escaped, let it return `Markup` strings.

File: hasjob/templates/stats.html.jinja2 [hunk 7]
```diff
@@ -21,7 +21,7 @@
       </div>
     </div>
 {% endblock %}
-{% block footerscripts %}
+{% block sheetscripts %}
```
Comment (jace) [inline hasjob/templates/stats.html.jinja2]: These templates should continue to use `footerscripts`. Only overwrite `sheetscripts` or `tablayoutscripts` in layout templates.

File: hasjob/templates/tablayout.html.jinja2 [hunk 8]
```diff
@@ -27,8 +27,11 @@
 {% endblock %}
 
 {% block layoutscripts %}
-  {%- if header_campaign %}{{ campaign_script() }}{% endif %}
+  {{ super() }}
   <script type="text/javascript" src="//cdnjs.cloudflare.com/ajax/libs/d3/3.5.17/d3.min.js"></script>
   <script type="text/javascript" src="//cdnjs.cloudflare.com/ajax/libs/c3/0.4.14/c3.min.js"></script>
-  {% block footerscripts %}{% endblock %}
-{% endblock %}
+  {% block tablelayoutscripts %}
```
Comment (jace) [inline hasjob/templates/tablayout.html.jinja2]: `{% block tablayoutscripts %}`

File: hasjob/views/index.py [hunk 9]
```diff
@@ -776,3 +776,18 @@ def logoimage(domain, hashid):
...
+
+
+@app.route('/service-worker.js', methods=['GET'])
+def sw():
+    return app.send_static_file('service-worker.js')
+
+
+@app.route('/workbox-sw.prod.v2.1.2.js', methods=['GET'])
+def workbox():
+    return app.send_static_file('workbox-sw.prod.v2.1.2.js')
```
Comment (jace) [inline hasjob/views/index.py]: What is `v2.1.2`?

File: hasjob/templates/stats.html.jinja2 (hunk 7 above)
Comment (vidya-ram) [reply hasjob/templates/stats.html.jinja2]: In `stats.html.jinja`
  `{% block sheetscripts %}`
      `{% block footerscripts %}`
            .....
      `{% endblock %}`
  `{% endblock %}`
Would this be fine, @jace ?

File: hasjob/views/index.py (hunk 9 above)
Comment (vidya-ram) [reply hasjob/views/index.py]: `workbox-webpack-plugin` generates the service worker that import's Workbox's service worker library
https://developers.google.com/web/tools/workbox/get-started/webpack#register

File: hasjob/templates/tablayout.html.jinja2 (hunk 8 above)
Comment (vidya-ram) [reply hasjob/templates/tablayout.html.jinja2]: fixed

File: hasjob/templates/layout.html.jinja2 (hunk 6 above)
Comment (vidya-ram) [reply hasjob/templates/layout.html.jinja2]: Changed

File: hasjob/templates/admin_dashboard.html.jinja2 (hunk 5 above)
Comment (vidya-ram) [reply hasjob/templates/admin_dashboard.html.jinja2]: Fixed

File: hasjob/static/sass/_header.sass (hunk 4 above)
Comment (vidya-ram) [reply hasjob/static/sass/_header.sass]: Fixed

File: hasjob/static/js/styles-app.js (hunk 3 above)
Comment (vidya-ram) [reply hasjob/static/js/styles-app.js]: Fixed

File: hasjob/__init__.py (hunk 2 above)
Comment (vidya-ram) [reply hasjob/__init__.py]: Removed

File: hasjob/templates/stats.html.jinja2 (hunk 7 above)
Comment (jace) [reply hasjob/templates/stats.html.jinja2]: Using just `{% block footerscripts %}` should work.

File: hasjob/views/index.py (hunk 9 above)
Comment (jace) [reply hasjob/views/index.py]: So v2.1.2 is a static path that won't change? Or it's linked to the webpack version? If latter, we should avoid a dependency on having to change this line every time webpack changes.

File: hasjob/templates/stats.html.jinja2 (hunk 7 above)
Comment (vidya-ram) [reply hasjob/templates/stats.html.jinja2]: In `sheet.html.jinja2`, I have added 
`{% block layoutscripts %}`
  `{{ super() }}`
`{% endblock %}`
And used just `{% block footerscripts %}` in templates that extend `sheet.html.jinja2`

File: hasjob/templates/sheet.html.jinja2 [hunk 10]
```diff
@@ -22,6 +22,5 @@
 {% endblock %}
 
 {% block layoutscripts %}
-  {%- if header_campaign %}{{ campaign_script() }}{% endif %}
-  {% block footerscripts %}{% endblock %}
+  {{ super() }}
 {% endblock %}
```
Comment (jace) [inline hasjob/templates/sheet.html.jinja2]: We don't need to define a block if it _only_ contains `super()`.

File: hasjob/views/index.py [hunk 11]
```diff
@@ -819,3 +819,18 @@ def logoimage(domain, hashid):
 @app.route('/search')
 def search():
     return redirect(url_for('index', **request.args))
+
+
+@app.route('/offline')
```
Comment (jace) [inline hasjob/views/index.py]: Can we be careful about adding new top-level routes? This merits a discussion. See #145.

**Comment by jace** [comment]
I have just these two comments, @vidya-ram.

File: hasjob/templates/sheet.html.jinja2 (hunk 10 above)
Comment (vidya-ram) [reply hasjob/templates/sheet.html.jinja2]: Removed.

File: hasjob/views/index.py (hunk 11 above)
Comment (vidya-ram) [reply hasjob/views/index.py]: Changed it to `/api/1/template/offline`

**Pull request merged**